docs/decisions/07-posts-data-model.md

```diff
@@ -0,0 +1,162 @@
+        string summary
+        string preview
+        string source
+ }
+    
+```
+
+So, everything becomes a post, with some child content items. It doesn't currently happen (much?) in the wild, but this would allow us to easily handle activities that include multiple content-like objects (multiple notes, images, questions, mix-and-match, etc). The types above are mostly illustrative, not complete or final. The idea is that they would implement an interface, call it `IContent`, which would provide enough information to build a feed of posts. That would make it a lot easier to actually provide those feeds to clients. Without *something* like that, clients would have had to make at least one follow-up request to get the actual content of feed items that were otherwise only a list of IDs/URLs.
+
+```mermaid
+classDiagram
+    direction LR
+    class IContent {
+        <<interface>>
+        Uuid Id
+        Uri Uri
+        Uri Thread
+        string? Summary
+        string? Preview
+        string? Source
+        
+        Type() string
+    }
+    
+    class Post {
+        Uuid Id
+        Uri Uri
+        Uri Thread
+        DateTime Created
+        DateTime Updated
+        IList~IContent~ Contents
+        string Client
+    }
+    
+    class Note {
+        string Content
+        Type() "Note"
+    }
+
+    class Image {
+        MediaType MediaType
+        Type() "Image"
+    }
+    
+    class FeedDto {
+        Post Post
+        Profile CreatedBy
+        Profile SharedBy
+        Post InReplyTo
+        Uri Thread
+        Content Content
+    }
+    
+    class Content {
+        Uuid Id
+        Uri Uri
+        Uri Thread
+        string? Summary
+        string? Preview
+        string? Source
+        
+        Type()*
+    }
+
+    Content <|-- Note : inherits
+    Content <|-- Image : inherits
+    Content <|-- Etc : inherits
+    IContent <|-- Content : implements
+    Post --* IContent : contains
+    Post --> FeedDto : maps-to
+    FeedDto --* Content : contains
+```
+
+## Impact
+
+This will obviously introduce some new data types. It's also likely to require some complex migrations. It should ideally happen soon, before any serious work starts on posting-related features.
+
+You may also notice that this data model is more restrictive than what ActivityPub would allow. Notably, Posts can only be InReplyTo at most one other Post (and no other kind of object). A Note can only be in one Post and vice-versa. And because `IContent.Thread` would map to `AS Context`, a post can only have one context and exist in one thread. These restrictions should be fully compatible with the actual behavior of other fedi services in the wild. It helps to narrow the possibility space that we would actually have to deal with, and should support faster queries for the most common cases (when a post is a note and nothing else). This means that there could be potential future fedi peers that we would not be readily compatible with.
```

Comment:
> A Note can only be in one Post and vice-versa.

Does this align with the discussion below about a content item theoretically existing in multiple posts?

Reply:
I think so. I meant for it to. The ADR is mostly about the internal data model. But this specific paragraph is about activitypub messages, and how this theoretically narrows the message compositions we can understand. But I don't think any existing implementation is actually doing any of the things this would exclude, so it's really just theoretical.